### DISCUSSION TITLE
How do I drive OrcaSlicer from command line?

### DISCUSSION THREAD

**Discussion by vaaronna** [OrcaSlicer/OrcaSlicer#1603] [Q&A]
I'd like to drive OrcaSlicer from the command line so I can automate some tests.  Can someone who has done this share some examples, nothing I'm trying seems to work.

**Comment by andronick83** [comment]
I was also looking for a manual.
The command line displays this:
```
$ path-to/orca-slicer[.exe] -h

OrcaSlicer-1.8.0-beta2:
Usage: orca-slicer [ OPTIONS ] [ file.3mf/file.stl ... ]

OPTIONS:
 --datadir           Load and store settings at the given directory. This is useful for maintaining
                     different profiles or including configurations from a network storage.
 --debug level       Sets debug logging level. 0:fatal, 1:error, 2:warning, 3:info, 4:debug, 5:trace
 --load-custom-gcodes custom_gcode_toolchange.json
                     Load custom gcode from json
 --load-filaments "filament1.json;filament2.json;..."
                     Load filament settings from the specified file list
 --load-settings "setting1.json;setting2.json"
                     Load process/machine settings from the specified file
 --outputdir dir     Output directory for the exported files.
 --skip-objects "3,5,10,77"
                     Skip some objects in this print
 --uptodate-settings "setting1.json;setting2.json"
                     load uptodate process/machine settings from the specified file when using
                     uptodate
 --arrange option    Arrange options: 0-disable, 1-enable, others-auto
 --convert-unit      Convert the units of model
 --ensure-on-bed     Lift the object above the bed when it is partially below. Disabled by default
 --orient            Orient options: 0-disable, 1-enable, others-auto
 --repetitions count Repetions count of the whole model
 --rotate            Rotation angle around the Z axis in degrees.
 --rotate-x          Rotation angle around the X axis in degrees.
 --rotate-y          Rotation angle around the Y axis in degrees.
 --scale factor      Scale the model by a float factor
 --export-3mf filename.3mf
                     Export project as 3MF.
 --export-settings settings.json
                     Export settings to a file.
 --export-slicedata slicing_data_directory
                     Export slicing data to a folder.
 --export-stl        Export the objects as multiple STL.
 --help, -h          Show command help.
 --info              Output the model's information.
 --load-defaultfila option
                     Load first filament as default for those not loaded
 --load-slicedata slicing_data_directory
                     Load cached slicing data from directory
 --min-save option   export 3mf with minimum size.
 --mstpp time        max slicing time per plate in seconds.
 --mtcpp count       max triangle count per plate for slicing.
 --no-check option   Do not run any validity checks, such as gcode path conflicts check.
 --normative-check option
                     Check the normative items.
 --pipe pipename     Send progress to pipe.
 --slice option      Slice the plates: 0-all plates, i-plate i, others-invalid
 --uptodate          Update the configs values of 3mf to latest.

Print settings priorites:
        1) setting values from the command line (highest priority)
        2) setting values loaded with --load_settings and --load_filaments
        3) setting values loaded from 3mf(lowest priority)
```

I learned that you can launch a ready-made 3mf project and immediately add the desired “stl” object to it:
```
path-to/orca-slicer[.exe] path-to/project.3mf path-to/object.stl
```

I have not yet found out whether it is possible to immediately center an object.

**Reply by andronick83** [reply]
By the way, in the latest version the help output is broken. And the version it displays is incorrect:
```
$ OrcaSlicer_Windows_V1.9.0-alpha_portable/OrcaSlicer/orca-slicer.exe -h --debug 5

[2023-12-29 16:39:38.063693] [0x000012b4] [trace]   Initializing StaticPrintConfigs
[2023-12-29 16:39:38.077726] [0x000012b4] [warning] cli mode, Current OrcaSlicer Version 01.08.02.56
[2023-12-29 16:39:38.077726] [0x000012b4] [info]    Will start to read model file now, file count :0

[2023-12-29 16:39:38.077726] [0x000012b4] [info]    allow_multicolor_oneplate 1, allow_rotations 1 skip_modified_gcodes 0 avoid_extrusion_cali_region 0 loaded_filament_ids size 0, clone_objects size 0
[2023-12-29 16:39:38.077726] [0x000012b4] [info]    plate_to_slice=0, normative_check=1, use_first_fila_as_default=0
[2023-12-29 16:39:38.077726] [0x000012b4] [info]    Slic3r::CLI::run:before load settings, file count=0

[2023-12-29 16:39:38.077726] [0x000012b4] [info]    current printer , new printer , current process , new process
[2023-12-29 16:39:38.078227] [0x000012b4] [info]    current printer inherits , new printer inherits , current process inherits , new process inherits
[2023-12-29 16:39:38.078227] [0x000012b4] [info]    old 3mf, no compatible printers, set to compatible
[2023-12-29 16:39:38.078227] [0x000012b4] [info]    old printer , inherited from , old process , inherited from  ,compatible 1
[2023-12-29 16:39:38.078227] [0x000012b4] [warning] no filament colors found in projects
[2023-12-29 16:39:38.078227] [0x000012b4] [info]    total 0 models, 0 objects

[2023-12-29 16:39:38.078691] [0x000012b4] [debug]   Slic3r::GUI::PartPlate::set_index: plate_id update from 0 to 0
[2023-12-29 16:39:38.078691] [0x000012b4] [debug]   Slic3r::GUI::PartPlate::set_index: plate_id update from 0 to 1
Segmentation fault
```

**Reply by vaaronna** [reply]
All the CLI info and development is in the upstream Bambuslicer, this isn't a good place to ask for help about the CLI.   You should file a bug report on that segfault though.

**Reply by vaaronna** [reply]
BS has been centering my CLI stl's automatically, although I did see it once into using absolute coordinates (probably a bug) from the GUI when I imported a 3mf (not opened, imported) and selected no when it asked to combine the objects.

**Reply by fg89o** [reply]
I'm trying to drive orca slicer 2.3.0 from command line and i can't get it working. Command -h and --help doesn't work. Where can I found the commands available in each version?. Thank you!

**Comment by charely6** [comment]
so assuming this is suppose to be similiar to Prusaslicer (slic3r originally) cli interface the --export-stl option doesn't seem to do anything with PrusaSlicer I can just do 
`"C:\Program Files\Prusa3D\PrusaSlicer\prusa-slicer.exe" --export-stl "Head Angry.3mf"`
and it will create a stl of that 3mf file while the orcaslicer version seems to do nothing?

I noticed this was part of slic3r from their documentation and I had hoped it was still part of orca-slicer so I could use it for a script based 3mf-> stl converter.

**Comment by vaaronna** [comment]
@charely6 since you already have a working "3mf to stl converter" in prusa-slicer.exe, why not just use that executable in your script if it is doing what you want?

**Reply by charely6** [reply]
I was hoping to be able to use orcaslicer because that's what I'm using for my primary slicer and I just shared the lazy batch script for converting 3mf to stl files and I wanted it so anyone who wanted to use it and only had orca slicer installed it could use that.

**Reply by vaaronna** [reply]
What is the use case?  It doesn't really seem to do anything useful unless maybe you have multiple objects, in that case does it output a stl with everything as they are laid out by the plater?  I don't see the utility in that either.  The 3mf already contains the mesh geometry, for a single object this conversion is trivial and there are probably several tools that do it already, you could write it in python in an hour.  If you are bulk converting 3mf's to STL's I wouldn't use a slicer to do it.

**Reply by charely6** [reply]
so it didn't seem to work for multiple objects. 
The reason I wanted it is I've had some weird issues when I import a downloaded 3mf load it up and then add other objects to be printed and OrcaSlicer saves that build plate of various objects over the original 3mf file. I know this is a pretty niche issue and I think I found a setting in OrcaSlicer to turn this off but I haven't really tested it. 
I thought there would be other tools as well but I couldn't find any that were just offline tools to do it. In my googling the answers I was finder were either open in slicer and export or use an online service, I did find another github project of someone working on it but they don't have a stable release yet and I didn't really feel like trying to compile a whole thing for such a little thing. 
I know using the slicer in this way is kind of dumb because even trying to convert a dozen files took a minute or 2, but I didn't really need a quick way to do it just a coinvent way. I don't know how much I'll use it, but I like the option.

**Comment by Ale-Ponte** [comment]
Hi everyone,
I'm jumping back into this thread, even though some time has passed, to ask if there have been any updates regarding command-line usage.

I'm working on automating the slicing process to build an online quoting system: the idea is that a user uploads an .stl file through a web interface, selects parameters like material type, infill density, and quality, and the system automatically generates a .gcode file or, at least, estimates print time and material usage.

I've tried several commands listed in BambuStudio’s wiki (on Windows 10), but running them through CMD doesn’t produce any output files or useful feedback. Same issue with OrcaSlicer.

Has anyone had recent success or found alternative solutions?
Or perhaps external agents/scripts/workarounds that can automate these steps?

Thanks in advance to anyone willing to share updates!

**Reply by Mieston** [reply]
on windows, use Nightly Builds(2.3.1) instead of Stable Release(2.3.0), at least I can see -h works

**Reply by Mieston** [reply]
on windows, only -h works
use pyautogui to click gui or use cli on wsl

**Reply by Ale-Ponte** [reply]
Thanks for your feedback! 
To avoid more issues, I’ve decided to move the project to Linux where the CLI should behave more reliably.

Appreciate the help!

**Reply by phbone** [reply]
@Ale-Ponte Have you found if the slicing functionality works in Linux?

**Reply by Ale-Ponte** [reply]
Hey @phbone, not yet. I think I'll try next month.

**Reply by rishabhrkaushik** [reply]
Did anyone try this yet? I got -h to work with nightly build as well but not with stable release. How about running orca in docker first? That way I can also control the version of software being used.

**Reply by Mieston** [reply]
@rishabhrkaushik the easiest way may be use release appImage on linux. i have tested and it works. you can download different release version and call them for version control

**Reply by AFKFelix** [reply]
@rishabhrkaushik I tried running it in Docker, but I could never figure out how to do it. Did you manage to get it running in a container by any chance?

**Reply by rishabhrkaushik** [reply]
@AFKFelix 
I did get the docker working. No luck slicing it completely yet. Able to slice things partially if I give a particular format of 3mf but most stl/3mf are failing with some error or the other. Trying to figure out version incompatibilities. 

As to how to run the docker, 
```
docker run -d `
  --name orcaslicer `
  --security-opt seccomp=unconfined `
  -e TZ=America/Chicago `
  -p 4000:3000 `
  -p 3001:3001 `
  -v orcaslicer_config:/config `
  --restart unless-stopped `
  lscr.io/linuxserver/orcaslicer:latest
```

This worked for me. 

Slicing slightly worked for 
```
/opt/orcaslicer/bin/orca-slicer --allow-newer-file --avoid-extrusion-cali-region --no-check --slice
 0 --outputdir "/config/3Dmodels" "/config/3Dmodels/Part Studio 2 - Part 1.3mf"
```

but loading settings, slicing other models, stl files, 3mf files are still throwing a lot of errors. Need to get it working in stable condition. 

Any luck on your end?

**Reply by AFKFelix** [reply]
@rishabhrkaushik  
> @AFKFelix I did get the docker working. No luck slicing it completely yet. Able to slice things partially if I give a particular format of 3mf but most stl/3mf are failing with some error or the other. Trying to figure out version incompatibilities.
> 
> As to how to run the docker,
> 
> ```
> docker run -d `
>   --name orcaslicer `
>   --security-opt seccomp=unconfined `
>   -e TZ=America/Chicago `
>   -p 4000:3000 `
>   -p 3001:3001 `
>   -v orcaslicer_config:/config `
>   --restart unless-stopped `
>   lscr.io/linuxserver/orcaslicer:latest
> ```
> 
> This worked for me.
> 
> Slicing slightly worked for
> 
> ```
> /opt/orcaslicer/bin/orca-slicer --allow-newer-file --avoid-extrusion-cali-region --no-check --slice
>  0 --outputdir "/config/3Dmodels" "/config/3Dmodels/Part Studio 2 - Part 1.3mf"
> ```
> 
> but loading settings, slicing other models, stl files, 3mf files are still throwing a lot of errors. Need to get it working in stable condition.
> 
> Any luck on your end?

Thanks for getting back to me. I checked out this Docker image before, but it didn't meet all my requirements so I didn't continue with it. I am currently writing a REST API to slice models and the LinuxServer/OrcaSlicer container hosts the OrcaSlicer GUI as a web app, so its not what i need really. However it might be worth checking how they are running it in the background, so I think I will do this. 

I was trying to run it as an AppImage or using Flatpak inside the container, but neither seems to be possible. 

Would you mind sharing the errors you encountered when slicing with the method you used? I have some experience with the CLI in general, so maybe I can help.

**Reply by rishabhrkaushik** [reply]
@AFKFelix 
I'm working on similar requirement where I need API to slice the model. Then using Python, I'll do post processing on the gcode file before sending it to printer. 

My idea was, run orcaslicer on another docker so that I can have the version controlled version of orca running. That way it wouldn't interfere with my other orcaslicer which I use for other kinds of prints etc. I can upgrade that without breaking the system. Once that is done, my python API backend can run a CLI command and slice the model using orca command line utility, take the output and continue from there. 

I was able to get it partially running in a custom docker image extracting Appimage and using the extracted bin via cli (couldn't get GUI working on it yet) or even in the 3rd party image using kasmvnc where both GUI and CLI seems to work.

The problem I'm facing is slicing the model using the CLI. It gives errors which my presumption is due to version incompatibilities between GUI and CLI (don't know why) in linux appimage and flatpak. The errors are of different kinds, some saying the config files are not compatible between 2.3.0 and current version, similar with 3MF, if I use `--allow-newer-file` flag, then I get some error regarding triangulation, if I ignore that, then I get some gcode mismatch so it just seems a deep rabbit hole which might be solved by having the correct version. I'm able to slice the same on windows system though using the same file and command. 

When I do `orca-slicer --help` in this cli, 
I get
```
root@e1876566d4fd:~/config/Apps/squashfs-root/bin# ./orca-slicer --help
[2025-09-10 15:24:58.629368] [0x00007af115bff480] [trace]   Initializing StaticPrintConfigs
OrcaSlicer-01.10.01.50:
Usage: orca-slicer [ OPTIONS ] [ file.3mf/file.stl ... ]
```
GUI in kasmvnc does show V2.3.0
<img width="711" height="495" alt="image" src="https://github.com/user-attachments/assets/67821d0a-0194-40de-ba30-754ad920236e" />


When I do it in my windows system, I get
```
PS C:\Program Files\OrcaSlicer> OrcaSlicer-2.3.1-dev:
Usage: orca-slicer [ OPTIONS ] [ file.3mf/file.stl ... ]
```
GUI shows the same (2.3.1dev).

I don't know why there is this version mismatch though. 

What's your approach? Any feedback on my above analysis?

**Reply by AFKFelix** [reply]
@rishabhrkaushik 

After reading your comment, I checked the image again and managed to get my container working by extracting the AppImage. Thanks for the tip! Now that my API and OrcaSlicer are running in this container, I can slice without any problems using v2.3.0. The resulting GCode looks valid, but I haven't done a print from it yet. Just as a reference, my Dockerfile is [here](https://github.com/AFKFelix/orca-slicer-api/blob/main/Dockerfile) if you're interested.

The version issue in the CLI is really interesting. From my experience I always need to use the '--allow-newer-file' option, but this has been the case ever since I started using the CLI. But if I pass it, it seems to work fine.
If you look at the GCode sliced with the CLI, it shows the correct version (2.3.0 for me). I think the CLI just has the wrong version string (or smth like this) in the code because it doesn't really make sense to me that the slicing itself would be different between the CLI and the GUI. 
 It's interesting to see the CLI working on Windows btw, this wasn't working the last time I tried it.

**Reply by rishabhrkaushik** [reply]
@AFKFelix 
Can you send the cli command you're using to slice. STL file, settings file would also be useful. Maybe something is wrong with my files which would be interesting since I'm able to get it to work on GUI.

**Reply by AFKFelix** [reply]
@rishabhrkaushik 

Slicing depends on the user settings/API request for me, but here's an example of what it could look like inside the container: 
```bash
/app/squashfs-root/AppRun 
  --slice 0 
  --load-settings "/app/data/printers/printer.json;/app/data/presets/preset.json" 
  --load-filaments "/app/data/filaments/filament.json" 
  --allow-newer-file 
  --outputdir "/tmp/slice-XYZ/output" 
  /tmp/slice-XYZ/input/uploaded-file.stl
```

For me the profiles are uploaded to the data directory by the user and the STL file is included in the request and saved in the temporary input folder. The gcode and a `result.json` file will then be generated in the temporary output folder.

Here are my test profiles for slicing for the P1S. I can't add an STL file, but you can use a simple cube to test it, it should work fine.
[preset.json](https://github.com/user-attachments/files/22277756/preset.json)
[printer.json](https://github.com/user-attachments/files/22277760/printer.json)
[filament.json](https://github.com/user-attachments/files/22277757/filament.json)

You can also test my API with your file if you wish. You can find it here and run it in docker: https://github.com/AFKFelix/orca-slicer-api/tree/main

**Reply by rishabhrkaushik** [reply]
> @rishabhrkaushik 
> 
> Slicing depends on the user settings/API request for me, but here's an example of what it could look like inside the container: 
> ```bash
> /app/squashfs-root/AppRun 
>   --slice 0 
>   --load-settings "/app/data/printers/printer.json;/app/data/presets/preset.json" 
>   --load-filaments "/app/data/filaments/filament.json" 
>   --allow-newer-file 
>   --outputdir "/tmp/slice-XYZ/output" 
>   /tmp/slice-XYZ/input/uploaded-file.stl
> ```
> 
> For me the profiles are uploaded to the data directory by the user and the STL file is included in the request and saved in the temporary input folder. The gcode and a `result.json` file will then be generated in the temporary output folder.
> 
> Here are my test profiles for slicing for the P1S. I can't add an STL file, but you can use a simple cube to test it, it should work fine.
> [preset.json](https://github.com/user-attachments/files/22277756/preset.json)
> [printer.json](https://github.com/user-attachments/files/22277760/printer.json)
> [filament.json](https://github.com/user-attachments/files/22277757/filament.json)
> 
> You can also test my API with your file if you wish. You can find it here and run it in docker: https://github.com/AFKFelix/orca-slicer-api/tree/main
> 
Got it. I'm able to slice 3mf files but not STL files yet. Probably some settings mismatch. I'll check your files. Thanks.
